upgrades/upgrade_testing_framework/core/docker_framework_client.py

```diff
@@ -0,0 +1,115 @@
+import logging
+import docker.client
+from docker.errors import ImageNotFound, NotFound
+from docker.models.networks import Network
+from docker.models.containers import Container
+from docker.models.volumes import Volume
+
+
+class RemoveRunningContainerException(Exception):
+    def __init__(self, container: Container):
+        self.container = container
+        super().__init__("A running container should be stopped before attempting to remove: {}".format(container.id))
+
+
+class ExistingContainerException(Exception):
+    def __init__(self, container_name: str):
+        self.container_name = container_name
+        super().__init__("There already exists a container with the name: {}".format(container_name))
+
+
+class DockerFrameworkClient:
+
+    def __init__(self, logger=logging.getLogger(__name__), docker_client=docker.from_env()):
+        # TODO: Use a custom logger
+        self.logger = logger
+        self.docker_client = docker_client
+
+    def is_container_running(self, container: Container) -> bool:
+        # TODO: Make more robust, case where this attr is not valid
+        if container is None:
+            raise TypeError("Provided argument is not of type docker.models.containers.Container")
+        container_state = container.attrs['State']
+        return container_state == 'running'
+
+    def does_container_exist(self, container_name: str) -> bool:
```

Comment:
This feels like a method that probably shouldn't be invoked outside this module (too low-level).  As such, I'd prefix it with an underscore (`_`) to indicate it's "private".  This is merely a convention, but it's a hint to future maintainers and users that they probably shouldn't be invoking the method directly.

Reply:
Just an intuition, but I think the Framework code shouldn't be thinking about thinks like "containers" and should instead be focused on Nodes/Clusters.  The Docker side feels like the implementation rather than the abstraction, yeah?